framework/include/executioners/Executioner.h

```diff
@@ -153,6 +153,42 @@ class Executioner : public MooseObject,
    */
   static MooseEnum iterationMethods() { return MooseEnum("picard secant steffensen", "picard"); }
 
+  /**
+   * Get a general iteration number for the purpose of outputting, useful in the presence of a
+   * nested solve. This output iteration number may be set by the parent app for a sub-app. This
+   * behavior is decided by the Executioner/Executor/SolveObject in charge of the solve.
+   */
+  virtual unsigned int getIterationNumberOutput() const { return _output_iteration_number; }
```

Comment:
I wonder if this should take an execute flag key. Then we could think about nesting iteration number output, e.g. what iteration number am I on in my `EXEC_NONLINEAR` loop? And at the same time, what iteration number am I on in my `EXEC_LINEAR` loop? Or alternatively there is no key, but we return some kind of ordered map which is a sequence of execute flags and their iteration numbers

Reply:
>  Then we could think about nesting iteration number output, e.g. what iteration number am I on in my `EXEC_NONLINEAR` loop?

That's a type of behavior that opens up with optimization (or other outer iteration) type of problems. I think that that type of output is less relevant for optimization problems than the result of the forward-adjoint solve (the iteration number doesn't change as you solve physics problems, so you'd just end up with normal system-wide convergence output which can be obtained from a non-optimization solve). It also seems that nothing gets updated in the subapps (e.g. forward) on nonlinear or linear, even if the flags are set in the postprocessors/output blocks. I.e. I am getting no additional output as I enable linear/nonlinear "outer iteration" output on `CSV.C`.  

The flags we could set here from the optimization module (`OptimizeSolve`) are optimization-specific flags (`EXEC_FORWARD`, `EXEC_ADJOINT`, `EXEC_ADJOINT_TIMESTEP_BEGIN`,...).

Reply:
> linear/nonlinear "outer iteration"

Wouldn't these be "inner" iterations?

Reply:
Yes.

Reply:
My concern with the current API, which is of course a product of us pushing you to generalize, is that its unclear what `getIterationNumberOutput()` means on its own. Does it mean the time step number? The nonlinear iteration number? The linear iteration number? The adaptivity step in a grid refinement study of a steady solve? This is why I'd propose APIs like this

```c++
virtual const std::vector<std::pair<ExecFlagType, IterationInfo>> & getIterationInfo() const;
virtual const IterationInfo & getIterationNumber(const ExecFlagType &) const;
```

where `IterationInfo` is a structure whose data members we could brainstorm. Certainly a `step` or `iteration_number` would be one necessary data member.

What are your thoughts? Maybe we should have a Teams call with @GiudGiud and anyone else who would be interested. Perhaps @dschwen

Reply:
Using data structures to better describe the output is a good idea. Separately, but related, I think we could fall into similar "contradictions". For example, if `step` or `iteration_number` are data members: What kind of iteration is captured with the `iteration_number` member? Is it an optimization one (we don't want to say that explicitly in the framework)? It seems as if we want _special_ output for an application that, from the framework, we know nothing of. This type of special output would probably beg for child optimization module objects.

We can talk more offline. I can also use your `IterationInfo` approach above.

Reply:
The IterationInfo is paired with an execution flag in the APIs I proposed. So there would hopefully be no ambiguity. 

I could definitely see a future though where IterationInfo (or whatever it ended up being) is a class that can be derived from. I don’t know if the future is here yet

Reply:
An execution flag in the framework does not necessarily define the iteration type we are after. We are using here TIMESTEP_END because it works, but it can still be equivocal. I'm not opposed to doing that, but I believe the design issue wouldn't be resolved with that data structure.

Reply:
Applications can register their own execution flags as the optimization module does